timemachine/fe/dummy.py

```diff
@@ -0,0 +1,631 @@
+            anchors.append(a_idx)
+
+    return anchors
+
+
+def enumerate_anchor_groups(mol, core, dummy_group):
+    """
+    An anchor group is a set of core atoms that are allowed to interact with the dummy
+    atoms in a dummy group in a way that allows the partition function to be separated.
+    Unlike dummy groups, anchor groups do not partition the non-core atoms. They're allowed
+    to overlap core atom indices. In other words, anchor groups for any particular dummy
+    group is fully independent of other anchor groups for any other dummy group.
+
+    While the choices of atoms in an anchor group is arbitrary, one heuristic we use
+    to enumerate anchor groups is as follows. In order to determine:
+
+    1) the 1-2 bond anchor i, we find a core atom that is directly bonded a dummy atom
+    2) the 1-3 angle anchor j, we pick a core atom that is adjacent to i
+    3) the 1-4 torsion anchor k, we pick a core atom adjacent to j, but not equal to i.
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        Molecule of interest
+
+    core: list of int
+        core atoms
+
+    dummy_group: list of int
+        dummy atoms in a given group.
+
+    Returns
+    -------
+    tuple
+        Anchor groups of size 1, 2, and 3 atoms
+
+    """
+
+    anchors = set()
+
+    for dummy_atom in dummy_group:
+        for anchor_atom in identify_root_anchors(mol, core, dummy_atom):
+            anchors.add(anchor_atom)
+
+    l1s = []
+    l2s = []
+    l3s = []
+
+    for anchor in anchors:
+        l1, l2, l3 = identify_anchor_groups(mol, core, anchor)
+        l1s.extend(l1)
+        l2s.extend(l2)
+        l3s.extend(l3)
+
+    return l1s, l2s, l3s
+
+
+def identify_dummy_groups(mol, core):
+    """
+    A dummy group is a set of dummy atoms that are inserted or deleted in alchemical
+    free energy calculations. The bonded terms that involve dummy atoms need to be
+    judiciously pruned such that the partition function at the end-states remain
+    factorizable, and cancelleable. Dummy groups are subject to the following constraints:
+
+    1) They must not contain atoms in the core.
+    2) Dummy groups do not interact with other dummy groups.
+    3) Dummy groups interact with the core only through anchor atoms.
+
+    While the choices of dummy groups is arbitrary, this function partitions the dummy
+    atoms into multiple dummy groups using a heuristic:
+
+    1) Identify all 1-2, 1-3 edges between dummy atoms.
+    2) Generate an induced graph using edges from above set.
+    3) Disconnected components of the resulting induced graph are our dummy groups.
+
+    We limit ourselves to only 1-2 and 1-3 interactions in 1) because we're typically
+    guaranteed that there exists a direct bond or that they're at most one bond away.
+
+    For example:
+
+    ```
+    mol = Chem.MolFromSmiles("FC1CC1(F)N")
+    core = [1, 2, 3]
+    dg = identify_dummy_groups(mol, core)
+    assert_set_equality(dg, [{0}, {4, 5}])
+    ```
+
```

Comment:
Should parameters be documented here?

Reply:
Trying to not get too verbose for code examples - the documentation should be pretty clear here.